### PR TITLE
Add --skip-ui-tests parameter to tuist test command

### PR DESCRIPTION
**Pull request by mollyIV**
Resolves https://github.com/tuist/tuist/issues/2562

### Short description 📝

Add `--skip-ui-tests` parameter to `tuist test` command to skip UI Tests targets from being executed.

**The changes in a nutshell**

new project mapper called `PruneUITestsProjectMapper` has been introduced. It marks UI Tests targets as "to be pruned" (only if a `--skip-ui-tests` flag is passed) and then `CacheTreeShakingGraphMapper` removes marked targets ✨ 

Please let me know if you feel like more acceptance / unit tests should be added 🙏 

Kudos to @fortmarek for help 😊 

```bash
$ tuist test --skip-ui-tests
```

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase.
- [ ] The changes have been tested following the [documented guidelines](https://tuist.io/docs/contribution/testing-strategy/).
- [x] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [x] In case the PR introduces changes that affect users, the documentation has been updated.

### CODE REVIEW

File: Sources/TuistAutomation/ProjectMappers/PruneUITestsProjectMapper.swift [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import Foundation
+import TSCBasic
+import TuistCore
+import TuistGraph
+
+public final class PruneUITestsProjectMapper: ProjectMapping {
```
Comment (luispadron) [inline Sources/TuistAutomation/ProjectMappers/PruneUITestsProjectMapper.swift:6]: Maybe add some docs as to "what" kind of UI tests this mapper is pruning as its not immediately obvious.

File: Sources/TuistKit/Commands/TestCommand.swift [hunk 2]
```diff
@@ -46,6 +46,11 @@ struct TestCommand: ParsableCommand {
     )
     var configuration: String?
 
+    @Flag(
+        help: "When passed, it skips testing UI Tests targets."
+    )
+    var skipUiTests: Bool = false
```
Comment (luispadron) [inline Sources/TuistKit/Commands/TestCommand.swift:52]: Should we follow SPM here and use flag inversion, so: `--no-ui-tests` [here](https://github.com/apple/swift-argument-parser/blob/main/Documentation/02%20Arguments%2C%20Options%2C%20and%20Flags.md#using-flag-inversions-enumerations-and-counts). So we'd have `--ui-tests` enabled by default and then `--no-ui-tests` if users wish to not use it.

File: Sources/TuistKit/GraphMappers/AutomationGraphMapperProvider.swift [hunk 3]
```diff
@@ -22,6 +22,7 @@ final class AutomationGraphMapperProvider: GraphMapperProviding {
         mappers.append(
             TestsCacheGraphMapper(hashesCacheDirectory: testsCacheDirectory, config: config)
         )
+        mappers.append(CacheTreeShakingGraphMapper())
```
Comment (luispadron) [inline Sources/TuistKit/GraphMappers/AutomationGraphMapperProvider.swift:25]: Whats this cange for?

File: Sources/TuistKit/GraphMappers/AutomationProjectMapperProvider.swift [hunk 4]
```diff
@@ -37,6 +40,12 @@ final class AutomationProjectMapperProvider: ProjectMapperProviding {
             SourceRootPathProjectMapper()
         )
 
+        if skipUITests {
+            mappers.append(
+                PruneUITestsProjectMapper()
+            )
+        }
```
Comment (luispadron) [inline Sources/TuistKit/GraphMappers/AutomationProjectMapperProvider.swift:47]: Wonder if instead we should just always add the mapper and pass it the flag instead, the mapper would then decide what to prune and what not to prune?

File: Tests/TuistAutomationTests/ProjectMappers/PruneUITestsProjectMapperTests.swift [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        subject = nil
+        super.tearDown()
+    }
+
+    func test_prune_is_set_to_ui_targets() throws {
+        // Given
+        let project = Project.test(
+            targets: [
+                .test(name: "App", product: .app),
+                .test(name: "UITests", product: .uiTests),
```
Comment (luispadron) [inline Tests/TuistAutomationTests/ProjectMappers/PruneUITestsProjectMapperTests.swift:30]: Could we add another target here that is just `unitTest` so we can ensure we only filter uiTests

**Review by luispadron — COMMENTED** [review]
This is lookin' good! I think it will be helpful for users that are letting Tuist handle all the schemes & target generation.

File: Sources/TuistKit/Commands/TestCommand.swift (hunk 2 above)
Comment (luispadron) [reply Sources/TuistKit/Commands/TestCommand.swift:52]: Similar for `--no-unit-tests` with `--unit-tests` enabled by default

File: Sources/TuistAutomation/ProjectMappers/PruneUITestsProjectMapper.swift (hunk 1 above)
Comment (mollyIV) [reply Sources/TuistAutomation/ProjectMappers/PruneUITestsProjectMapper.swift:6]: Hmm, maybe renaming the mapper to `PruneUITestsTargetsProjectMapper` would express intent better? 

> "what" kind of UI tests this mapper is pruning

What about docs like:

```
// This mapper maks all UI Tests project's targets as "to be pruned".
public final class PruneUITestsProjectMapper: ProjectMapping {
```

I am opened to other suggestions.

File: Sources/TuistKit/GraphMappers/AutomationGraphMapperProvider.swift (hunk 3 above)
Comment (mollyIV) [reply Sources/TuistKit/GraphMappers/AutomationGraphMapperProvider.swift:25]: This mapper is getting rid of the targets marked as `prune`.

File: Sources/TuistKit/GraphMappers/AutomationProjectMapperProvider.swift (hunk 4 above)
Comment (mollyIV) [reply Sources/TuistKit/GraphMappers/AutomationProjectMapperProvider.swift:47]: I'd keep the mappers simple, meaning they always perform the logic and the client decides when to use them.

File: Tests/TuistAutomationTests/ProjectMappers/PruneUITestsProjectMapperTests.swift (hunk 5 above)
Comment (mollyIV) [reply Tests/TuistAutomationTests/ProjectMappers/PruneUITestsProjectMapperTests.swift:30]: Sure, will do 👍

File: Sources/TuistKit/Commands/TestCommand.swift (hunk 2 above)
Comment (mollyIV) [reply Sources/TuistKit/Commands/TestCommand.swift:52]: Can you please link the example of SPM command that uses this convention? By printing a help via `swift package --help` there's the flags like 

```
--skip-update           Skip updating dependencies from their remote during a resolution
```

IMHO `--skip-ui-tests` is easy to understand (no negation as it was with `no-ui-tests`) and it expresses intent clearly. 

Let see what others think. 

IMHO `--skip-ui-tests` flag is easy to read (no negation as it was with `--no-ui-tests`)

File: Package.resolved [hunk 6]
```diff
@@ -132,8 +132,8 @@
         "repositoryURL": "https://github.com/stencilproject/Stencil.git",
         "state": {
           "branch": null,
-          "revision": "94197b3adbbf926348ad8765476a158aa4e54f8a",
-          "version": "0.14.0"
+          "revision": "973e190edf5d09274e4a6bc2e636c86899ed84c3",
```
Comment (pepibumur) [inline Package.resolved:135]: This seems unrelated to the PR. What about doing it in a separate PR?

File: Sources/TuistAutomation/ProjectMappers/PruneUITestsProjectMapper.swift (hunk 1 above)
Comment (pepibumur) [reply Sources/TuistAutomation/ProjectMappers/PruneUITestsProjectMapper.swift:6]: I agree with Luis here. I think it's important to capture the "intent" of this mapper. Why would a user prune the UI targets from the project. Since the goal is to "skip certain test targets", what about including `skip` in the name? For example, `SkipUITestsProjectMapper`. I think the concept of "pruning" is an internal implementation detail that I don't think should be captured in the name.

File: Sources/TuistAutomation/ProjectMappers/PruneUITestsProjectMapper.swift [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    public init() {}
+
+    public func map(project: Project) throws -> (Project, [SideEffectDescriptor]) {
```
Comment (pepibumur) [inline Sources/TuistAutomation/ProjectMappers/PruneUITestsProjectMapper.swift:9]: I think we should revisit the idea of "pruning". I think long-term, it's better if mappers just do what they are designed for. In this case removing certain targets for the project. This might leave the project in a invalid state, for example schemes with broken references to targets. For that reason I think we should extract from `CacheTreeShakingGraphMapper` the logic that ensures the integrity of the project. We can place that in a `IntegrityGraphMapper`. @luispadron do you think "integrity" is the right word for describe what it does?

File: Sources/TuistKit/Commands/TestCommand.swift (hunk 2 above)
Comment (pepibumur) [reply Sources/TuistKit/Commands/TestCommand.swift:52]: I'm personally inclined towards `--skip-ui-tests` because `skip` captures the intent of the flag, where as `no` is a bit ambiguous.

File: Sources/TuistKit/Commands/TestCommand.swift [hunk 8]
```diff
@@ -46,6 +46,11 @@ struct TestCommand: ParsableCommand {
     )
     var configuration: String?
 
+    @Flag(
```
Comment (pepibumur) [inline Sources/TuistKit/Commands/TestCommand.swift:49]: Nitpick: I'd extend the annotation to be more explicit about the short version of the flag. Since it's UI tests, what about making it `-u`?

File: Sources/TuistKit/GraphMappers/AutomationGraphMapperProvider.swift (hunk 3 above)
Comment (pepibumur) [reply Sources/TuistKit/GraphMappers/AutomationGraphMapperProvider.swift:25]: As I mentioned in one of the comments, I'd extract the logic of ensuring the project integrity into a new mapper and use that one instead.

File: Sources/TuistKit/GraphMappers/AutomationProjectMapperProvider.swift (hunk 4 above)
Comment (pepibumur) [reply Sources/TuistKit/GraphMappers/AutomationProjectMapperProvider.swift:47]: I'm inclined towards @mollyIV's suggestion here. The pattern that we've followed so far is letting the provider add/remove mappers depending on the flags and the project configuration. It's been working fine so I'd be consistent and not reconsider that as part of this PR.

**Review by pepibumur — CHANGES_REQUESTED** [review]
This is great @mollyIV. From my suggestions the one I'd try to address is removing the "prune" thing and have a mapper that ensures the integrity of the project. This integrity mapper should always be the last one.

**Comment by pepibumur** [comment]
I forgot to mention that we should update the documentation to document what the new flag is for. Moreover, I'd consider adding an acceptance test that verifies that UI tests are skipped. You can set up the acceptance test to run `tuist test` and verify from the list of run tests whether the UI ones were skipped.

**Review by fortmarek — COMMENTED** [review]
Looking good! Thanks @mollyIV 👏

File: Sources/TuistKit/GraphMappers/AutomationProjectMapperProvider.swift (hunk 4 above)
Comment (luispadron) [reply Sources/TuistKit/GraphMappers/AutomationProjectMapperProvider.swift:47]: Sounds good, thanks for clarifying!

File: Sources/TuistKit/Commands/TestCommand.swift (hunk 8 above)
Comment (mollyIV) [reply Sources/TuistKit/Commands/TestCommand.swift:49]: Hmm, if at some point we decided to add `skip-unit-tests` option, wouldn't a `-u` short version be confusing? 🤔 Maybe `-ui`?

File: Sources/TuistAutomation/ProjectMappers/PruneUITestsProjectMapper.swift (hunk 7 above)
Comment (mollyIV) [reply Sources/TuistAutomation/ProjectMappers/PruneUITestsProjectMapper.swift]: I like the idea, however maybe we can tackle that in the following PR from two reasons:
* I'd like to keep the size of a pull request small 
* getting rid of puring feels out of scope for this PR 

I have already been playing with extracting the integrity part from `CacheTreeShakingGraphMapper`. I'd like to write down a quick summary to be sure we are on the same page 🙏  

1. `SkipUITestsProjectMapper` would remove the ui tests targets explicitly instead of marking them as to be pruned. 
```swift
public final class SkipUITestsProjectMapper: ProjectMapping {
    public func map(project: Project) throws -> (Project, [SideEffectDescriptor]) {
        var project = project
        project.targets = project.targets.filter { $0.product != .uiTests }
        return (project, [])
    }
}
```

2. `IntegrityGraphMapper` implementation would be identical to the current state of `CacheTreeShakingGraphMapper`, but it would do nothing related to `target.prune`, meaning this would be removed: 
https://github.com/tuist/tuist/blob/40982d094e61247c263670d86c0140e77eb76619/Sources/TuistCache/GraphMappers/CacheTreeShakingGraphMapper.swift#L13

3. We'd need to add `IntegrityGraphMapper` as a last mapper to every place where the `CacheTreeShakingGraphMapper` is added. 

4. `CacheGraphMutator` sets targets to be pruned - how we would migrate this part? Should the mutator remove the targets directly? Would we need `CacheTreeShakingGraphMapper` then still? I am a bit confused how those would fit together. 

Thanks for the feedback!

Comment (mollyIV) [reply Sources/TuistAutomation/ProjectMappers/PruneUITestsProjectMapper.swift]: Hope you do not mind, but I have created a follow-up ticket to tackle pruning refactoring 🙇 

* https://github.com/tuist/tuist/issues/3078

**Comment by mollyIV** [comment]
> I forgot to mention that we should update the documentation to document what the new flag is for. 

The documentation has been updated 👌 

> Moreover, I'd consider adding an acceptance test that verifies that UI tests are skipped. You can set up the acceptance test to run tuist test and verify from the list of run tests whether the UI ones were skipped.

TBH I'd stick to unit tests only as most of the logic is contained.

**Comment by mollyIV** [comment]
Hey @pepibumur , @fortmarek, @luispadron  👋 

Thanks a lot for your feedback ❤️ I believe every comment / suggestion has been addressed and the pull request is ready for re-review 👀

**Review by pepibumur — APPROVED** [review]

File: projects/docs/docs/commands/test.md [hunk 9]
```diff
@@ -57,3 +57,4 @@ One of the benefits of using Tuist over other automation tools is that developer
 | `--device`          | `-d`  | `Test on a specific device.`                                        |                   | No       |
 | `--os`              | `-o`  | `Test with a specific version of the OS.`                           |                   | No       |
 | `--configuration`   | `-C`  | `The configuration to be used when building the scheme.`            |                   | No       |
+| `--skip-ui-tests`   | `-C`  | `When passed, it skips testing UI Tests targets.`                   | False             | No       |
```
Comment (fortmarek) [inline projects/docs/docs/commands/test.md:60]: I believe the short flag here is wrong.

File: Tests/TuistKitTests/Services/TestServiceTests.swift [hunk 10]
```diff
@@ -265,15 +265,17 @@ private extension TestService {
         configuration: String? = nil,
         path: AbsolutePath,
         deviceName: String? = nil,
-        osVersion: String? = nil
+        osVersion: String? = nil,
+        skipUiTests: Bool = false
```
Comment (fortmarek) [inline Tests/TuistKitTests/Services/TestServiceTests.swift:269]: Nit: There is inconsistent usage of `skipUiTests` and `skipUITests` - can you just do a quick search and replace, so only `skipUITests` is used?

**Review by fortmarek — APPROVED** [review]
Some minor comments, but otherwise good to merge IMHO 👏

File: projects/docs/docs/commands/test.md (hunk 9 above)
Comment (mollyIV) [reply projects/docs/docs/commands/test.md]: Fixed, by not allowing a short. The reason behind it mentioned [here](https://github.com/tuist/tuist/pull/2832#discussion_r631565504).

File: Tests/TuistKitTests/Services/TestServiceTests.swift (hunk 10 above)
Comment (mollyIV) [reply Tests/TuistKitTests/Services/TestServiceTests.swift:269]: Fixed 💪
